[Public Preview] Azure Confidential Ledger

lynshi opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

<!-- Please add a brief description of the changes made in this PR -->

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated: (N / A, preview version)
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1135
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
(automatically generated under `docs/` as part of the autorest workflow)

dolauli commented:
/azp run

BethanyZhou commented on src/ConfidentialLedger/README.md:
```diff
@@ -0,0 +1,64 @@
+# You need to specify your swagger files here.
+  # - $(repo)/specification/databricks/resource-manager/Microsoft.Databricks/stable/2018-04-01/databricks.json
+# If the swagger has not been put in the repo, you may uncomment the following line and refer to it locally
+  - $(this-folder)/../../../azure-rest-api-specs/specification/confidentialledger/resource-manager/Microsoft.ConfidentialLedger/preview/2021-05-13-preview/confidentialledger.json
```
Hi @lynshi , please place the swagger path with a public available link. Here is an example for you: https://github.com/Azure/azure-powershell/blob/03527c95d8ad02611bfa7eca0fbacf1583ece6c0/src/ContainerInstance/README.md?plain=1#L39

lynshi replied:
Sorry, the latest version wasn't available when I was working on this. I've updated the link now.

BethanyZhou reviewed (commented):
Please submit a design review for your cmdlets.

BethanyZhou commented on src/ConfidentialLedger/examples/Get-AzConfidentialLedger.md:
```diff
@@ -0,0 +1,20 @@
+
+Location Name               
+eastus   testledger0
+eastus   testledger1
```
It should follow the schema below :

```powershell 
PS C:\> {{ Add code here }}
{{ Add output here }}
```

{{ Add description here }}

Please add description for all examples

BethanyZhou commented on src/ConfidentialLedger/examples/New-AzConfidentialLedger.md:
```diff
@@ -0,0 +1,25 @@
+      } `
+  -CertBasedSecurityPrincipal `
+      @{
+          Cert="-----BEGIN CERTIFICATE-----MIIBsjCCATigAwIBAgIUZWIbyG79TniQLd2UxJuU74tqrKcwCgYIKoZIzj0EAwMwEDEOMAwGA1UEAwwFdXNlcjAwHhcNMjEwMzE2MTgwNjExWhcNMjIwMzE2MTgwNjExWjAQMQ4wDAYDVQQDDAV1c2VyMDB2MBAGByqGSM49AgEGBSuBBAAiA2IABBiWSo/j8EFit7aUMm5lF+lUmCu+IgfnpFD+7QMgLKtxRJ3aGSqgS/GpqcYVGddnODtSarNE/HyGKUFUolLPQ5ybHcouUk0kyfA7XMeSoUA4lBz63Wha8wmXo+NdBRo39qNTMFEwHQYDVR0OBBYEFPtuhrwgGjDFHeUUT4nGsXaZ ...
```
CERTIFICATE should be hidden otherwise we aren't able to pass security check.

lynshi replied:
This is a public key and we include it in the example for all of our other public releases too (Swagger, Azure CLI, etc). Will your CI specifically reject this, or is it ok to keep it here?

BethanyZhou replied:
It's risky. I would like to suggest you to represent the key with a obscure asterisk， like 
```
          Cert="-----BEGIN CERTIFICATE-----**************************-----END CERTIFICATE-----"; 
```

lynshi replied:
I've hidden it in the examples and docs. However, it is still part of the tests and recordings, and it has to be there because we validate the certificate in the resource provider.

Just wanted to reiterate that this certificate is intended to be public information and has no security value.

BethanyZhou replied:
Got it. I just want to hide it from docs to avoid security risk alert.

lynshi commented:
I'm not sure if I still need to generate the help files as described in the checklist. They didn't seem to generate correctly (the examples were missing), and I think the new auto code generation automatically generates help under the `docs` folder already.

lynshi commented:
I've covered the issues from the design review, but there's a few problems with the newly autogenerated code. I think a newer version for `autorest` may have been released since the last time I did this, and the previous recordings are no longer passing.

BethanyZhou commented:
Hi @lynshi ,

there are two issues reported in CI.

- D:\a\1\s\artifacts//StaticAnalysisResults\SignatureIssues.csv Errors
  "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
  "Az.ConfidentialLedger","New-AzConfidentialLedgerAADBasedSecurityPrincipalObject","New-AzConfidentialLedgerAADBasedSecurityPrincipalObject","1","8100","New-AzConfidentialLedgerAADBasedSecurityPrincipalObject Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
  "Az.ConfidentialLedger","New-AzConfidentialLedgerCertBasedSecurityPrincipalObject","New-AzConfidentialLedgerCertBasedSecurityPrincipalObject","1","8100","New-AzConfidentialLedgerCertBasedSecurityPrincipalObject Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should implement Force / ShouldContinue"
  
They are false positive. Please suppress them by adding SignatureIssues.csv in tools/StaticAnalysis/Exceptions/Az.ConfidentialLedger. Here is an example for you: https://github.com/Azure/azure-powershell/blob/main/tools/StaticAnalysis/Exceptions/Az.Aks/SignatureIssues.csv

- Test in Playback mode reports "ParameterBindingValidationException: Cannot bind argument to parameter 'SubscriptionId' because it is null.". Please upload env.json file. It should be in the src/ConfidentialLedger/test  folder.

BethanyZhou commented on src/ConfidentialLedger/docs/Update-AzConfidentialLedger.md:
````diff
@@ -0,0 +1,321 @@
+```
+Update-AzConfidentialLedger -Name <String> -ResourceGroupName <String> [-SubscriptionId <String>]
+ [-AadBasedSecurityPrincipal <IAadBasedSecurityPrincipal[]>]
+ [-CertBasedSecurityPrincipal <ICertBasedSecurityPrincipal[]>] [-LedgerType <LedgerType>] [-Location <String>]
````
My mistake. I just find there is a parameter Location. I believe Location can't be updated. Is it true? Have you tested each parameter locally. We always meet the case like that. Some immutable parameter shows in update- command.

lynshi replied:
How can we hide it - is it a setting in the API swagger or the readme?

lynshi replied:
To clarify, it is a required field for the API call. We mark `location` with `x-mis-mutability: ["create", "read"]` in the [Swagger](https://github.com/Azure/azure-rest-api-specs/blob/06f446933347cf235bf85abe1afdbee49615bf9c/specification/confidentialledger/resource-manager/Microsoft.ConfidentialLedger/preview/2021-05-13-preview/confidentialledger.json#L415), which [indicates that the property can not be updated after creation](https://azure.github.io/autorest/extensions/#x-ms-mutability). How canI resolve this?

BethanyZhou replied:
Here is an instruction: https://github.com/Azure/autorest.powershell/blob/main/docs/directives.md#parameter-hiding. I will let xiaogang know this issue.

lynshi replied:
Just tested - our RP returns the following if no location is provided `"{\"error\":{\"code\":\"ResourcePatchValidateFailed\",\"message\":\"The resource validation failed.\"}}"`. I saw in the example that you can provide a default value via the module; however, I suspect that if a different value than the region in which the resource is located is passed in the request, the RP will also return an error. Thus, we cannot have the PowerShell module set a default value either.

BethanyZhou replied:
So I guess the location must be provided and it's one of identifying information. If so, I think we may need to mark the location as mandatory parameter so that customer have to provide it.

lynshi replied:
Is there a directive for that, or is it sufficient to just edit the documentation?

BethanyZhou replied:
It's insufficient. We need to change Location from optional to required in command syntax level. Customization 
can help us. There are 3 steps:
- Copy Update-AzConfidentialLedger.ps1 from exports folder to custom folder
- Hide generated Update-AzConfidentialLedger by directive
- Edit custom/Update-AzConfidentialLedger.ps1, set Location as mandatory
https://github.com/Azure/azure-powershell/blob/fd1c2fa8e72c911c8ac94d958cd89a62162b1a80/src/Kusto/custom/Update-AzKustoDataConnection.ps1#L179-L183

lynshi replied:
Got it, thanks! I've now hidden the auto-generated code and use the custom cmdlet.

lynshi commented:
Hi @BethanyZhou I've fixed some of the issues mentioned.

1. Added `SignatureIssues.csv`. Will take another look if the CI still fails.
2. Added new recordings based on the updated code generation. `pwsh ./test-module.ps1` succeeds locally based on the recording files.
3. Restored `env.json`. I had it in my `.gitignore` previously.

BethanyZhou reviewed (approved):
@lynshi Could you provide the change log message which will be used when we release this module?

BethanyZhou commented:
@lynshi

BethanyZhou commented on src/ConfidentialLedger/README.md:
```diff
@@ -0,0 +1,72 @@
+      verb: Test
+      variant: ^Check$|^CheckViaIdentity$|^CheckViaIdentityExpanded$
+    remove: true
+  - where: # Hide auto-generated Update since we implement a custom one requiring the 'Location' parameter.
+      variant: ^UpdateExpanded$|^UpdateViaIdentityExpanded$
+    hide: true
```
```suggestion
  - where: # Hide auto-generated Update since we implement a custom one requiring the 'Location' parameter.
      verb: Update
      Subject: ConfidentialLedger
    hide: true
```

BethanyZhou replied:
Sorry, I checked the command details for `Update-AzConfidentialLedger`

    /// [DETAILS]
    /// verb: Update
    /// subjectPrefix: Confidential
    /// subject: Ledger

Its subject should be Ledger

BethanyZhou commented on src/ConfidentialLedger/custom/Update-AzConfidentialLedger.ps1:
```diff
@@ -0,0 +1,244 @@
+
+begin {
+    try {
+        $outBuffer = $null
```
FYI. This implementation is fine, but usually we will invoke the hidden generated Update-AzConfidentialLedger to simplify code. Here is an example: https://github.com/Azure/azure-powershell/blob/main/src/ContainerInstance/custom/Invoke-AzContainerInstanceCommand.ps1#L130

lynshi commented:
> LGTM. Could you provide the change log message which will be used when we release this module?

Sorry, do you mean you want us to provide a CHANGELOG file for this module, or a message for the overall CHANGELOG for this repository? I had checked some other modules and noticed they didn't have a CHANGELOG.md until 1.0.0. Since this is a preview version, I'm not sure what we would write other than "Released preview version of the PowerShell module for Azure Confidential Ledger".

BethanyZhou commented:
> Sorry, do you mean you want us to provide a CHANGELOG file for this module, or a message for the overall CHANGELOG for this repository? I had checked some other modules and noticed they didn't have a CHANGELOG.md until 1.0.0. Since this is a preview version, I'm not sure what we would write other than "Released preview version of the PowerShell module for Azure Confidential Ledger".

Yes, I do need a changelog message for this module. Usually, we write it as `First preview release for module Az.xxx
` for onboarding module. Is this message acceptable for you? Please refer to https://github.com/Azure/azure-powershell/blob/main/src/DiskPool/Changelog.md

lynshi commented:
Sure, the message works for me! Please feel free to make the change if that helps you merge now.

I noticed you made a few additional suggestions. If those are blocking comments, I'll fix them tomorrow morning.

BethanyZhou commented:
No worries. I can take care of it. Have a good night.

BethanyZhou commented on src/ConfidentialLedger/custom/Update-AzConfidentialLedger.ps1:
```diff
@@ -0,0 +1,244 @@
+begin {
+    try {
+        $outBuffer = $null
+        if ($PSBoundParameters.TryGetValue('OutBuffer', [ref]$outBuffer)) {
+            $PSBoundParameters['OutBuffer'] = 1
+        }
+        $parameterSet = $PSCmdlet.ParameterSetName
+        $mapping = @{
+            UpdateExpanded = 'Az.ConfidentialLedger.private\Update-AzConfidentialLedger_UpdateExpanded';
+            UpdateViaIdentityExpanded = 'Az.ConfidentialLedger.private\Update-AzConfidentialLedger_UpdateViaIdentityExpanded';
+        }
+        if (('UpdateExpanded') -contains $parameterSet -and -not $PSBoundParameters.ContainsKey('SubscriptionId')) {
+            $PSBoundParameters['SubscriptionId'] = (Get-AzContext).Subscription.Id
+        }
+        $cmdInfo = Get-Command -Name $mapping[$parameterSet]
+        [Microsoft.Azure.PowerShell.Cmdlets.ConfidentialLedger.Runtime.MessageAttributeHelper]::ProcessCustomAttributesAtRuntime($cmdInfo, $MyInvocation, $parameterSet, $PSCmdlet)
+        $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand(($mapping[$parameterSet]), [System.Management.Automation.CommandTypes]::Cmdlet)
+        $scriptCmd = {& $wrappedCmd @PSBoundParameters}
+        $steppablePipeline = $scriptCmd.GetSteppablePipeline($MyInvocation.CommandOrigin)
+        $steppablePipeline.Begin($PSCmdlet)
+    } catch {
+        throw
+    }
+}
+
+process {
+    try {
+        $steppablePipeline.Process($_)
+    } catch {
+        throw
+    }
+}
+
+end {
+    try {
+        $steppablePipeline.End()
+    } catch {
+        throw
+    }
+}
```
```suggestion
    process {
      try {
        Az.ContainerInstance.internal\Invoke-AzContainerInstanceCommand @PSBoundParameters
      } catch {
        throw
      }
    }
```

BethanyZhou commented on src/ConfidentialLedger/custom/Update-AzConfidentialLedger.ps1:
```diff
@@ -0,0 +1,211 @@
+
+    process {
+      try {
+        Az.ContainerInstance.internal\Invoke-AzContainerInstanceCommand @PSBoundParameters
```
```suggestion
        Az.ConfidentialLedger.internal\Update-AzConfidentialLedger @PSBoundParameters
```

BethanyZhou commented:
/azp run azure-powershell - powershell-core

BethanyZhou reviewed (approved):
LGTM

lynshi commented:
Thank you so much for your help @BethanyZhou!
